diff -rupN boost-1.53.0/boost/circular_buffer/base.hpp boost-1.53.0-patched/boost/circular_buffer/base.hpp
--- boost-1.53.0/boost/circular_buffer/base.hpp	2013-02-20 19:41:30.000000000 -0800
+++ boost-1.53.0-patched/boost/circular_buffer/base.hpp	2013-10-24 13:50:35.000000000 -0700
@@ -2087,7 +2087,7 @@ private:
             throw_exception(std::length_error("circular_buffer"));
 #if BOOST_CB_ENABLE_DEBUG
         pointer p = (n == 0) ? 0 : m_alloc.allocate(n, 0);
-        std::memset(p, cb_details::UNINITIALIZED, sizeof(value_type) * n);
+        std::memset((void*) p, cb_details::UNINITIALIZED, sizeof(value_type) * n);
         return p;
 #else
         return (n == 0) ? 0 : m_alloc.allocate(n, 0);
@@ -2130,7 +2130,7 @@ private:
         m_alloc.destroy(p);
 #if BOOST_CB_ENABLE_DEBUG
         invalidate_iterators(iterator(this, p));
-        std::memset(p, cb_details::UNINITIALIZED, sizeof(value_type));
+        std::memset((void*) p, cb_details::UNINITIALIZED, sizeof(value_type));
 #endif
     }
 
